nemo/core/connectors/save_restore_connector.py

```diff
@@ -0,0 +1,325 @@
+            else:
+                # artifact is optional and we simply return None
+                return None
+
+        assert os.path.exists(return_path)
+
+        artifact_item.path = os.path.abspath(src)
+        model.artifacts[config_path] = artifact_item
+        # we were called by ModelPT
+        if hasattr(model, "cfg"):
+            with open_dict(model._cfg):
+                OmegaConf.update(model.cfg, config_path, return_path)
+        return return_path
+
+    def _handle_artifacts(self, model, nemo_file_folder):
+        tarfile_artifacts = []
+        app_state = AppState()
+        for conf_path, artiitem in model.artifacts.items():
+            if artiitem.path_type == model_utils.ArtifactPathType.LOCAL_PATH:
+                if not os.path.exists(artiitem.path):
+                    raise FileNotFoundError(f"Artifact {conf_path} not found at location: {artiitem.path}")
+
+                # Generate new uniq artifact name and copy it to nemo_file_folder
+                # Note uuid.uuid4().hex is guaranteed to be 32 character long
+                artifact_base_name = os.path.basename(artiitem.path)
+                artifact_uniq_name = f"{uuid.uuid4().hex}_{artifact_base_name}"
+                shutil.copy2(artiitem.path, os.path.join(nemo_file_folder, artifact_uniq_name))
+
+                # Update artifacts registry
+                artiitem.hashed_path = "nemo:" + artifact_uniq_name
+                model.artifacts[conf_path] = artiitem
+
+            elif artiitem.path_type == model_utils.ArtifactPathType.TAR_PATH:
+                # process all tarfile artifacts in one go, so preserve key-value pair
+                tarfile_artifacts.append((conf_path, artiitem))
+
+            else:
+                raise ValueError(f"Directly referencing artifacts from other nemo files isn't supported yet")
+
+        # Process current tarfile artifacts by unpacking the previous tarfile and extract the artifacts
+        # that are currently required.
+        model_metadata = app_state.get_model_metadata_from_guid(model.model_guid)
+        if len(tarfile_artifacts) > 0 and model_metadata.restoration_path is not None:
+            # Need to step into nemo archive to extract file
+            # Get path where the command is executed - the artifacts will be "retrieved" there
+            # (original .nemo behavior)
+            cwd = os.getcwd()
+            try:
+                # Step into the nemo archive to try and find the file
+                with tempfile.TemporaryDirectory() as archive_dir:
+                    self._unpack_nemo_file(path2file=model_metadata.restoration_path, out_folder=archive_dir)
+                    os.chdir(archive_dir)
+                    for conf_path, artiitem in tarfile_artifacts:
+                        # Get basename and copy it to nemo_file_folder
+                        if 'nemo:' in artiitem.path:
+                            artifact_base_name = artiitem.path.split('nemo:')[1]
```

Comment:
hmmm, I haven't followed latest changes, but why you have this "nemo:" prefix  in artifact filename? seems you are getting rid of it anyway...

Reply:
This is infected into the saved model config, and reused to restore the hashed artifact from the tarfiles. The slice off is to get the hashed value, the key we look for is Nemo: